src/Backends/Firmware.vala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2022 elementary, Inc. (https://elementary.io)
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA
+ *
+ */
+
+public class SettingsDaemon.Backends.Firmware : GLib.Object {
+    private GLib.Settings firmware_settings;
+
+    private GLib.DateTime last_firmware_check = null;
+
+    private const int SECONDS_BETWEEN_REFRESHES = 60 * 60 * 24;
+
+    public uint firmware_updates_number { get; private set; default = 0U; }
+
+    private AsyncMutex update_notification_mutex = new AsyncMutex ();
+    private AsyncMutex firmware_update_notification_mutex = new AsyncMutex ();
+
+    public GLib.Application application { get; construct; }
+
+    public Firmware (GLib.Application application) {
+        Object (
+            application: application
+        );
+    }
+
+    construct {
+        firmware_settings = new GLib.Settings ("io.elementary.settings-daemon.firmware");
+
+        last_firmware_check = new DateTime.from_unix_utc (firmware_settings.get_int64 ("last-refresh-time"));
+
+        check_seconds_between_refreshes ();
+
+        Timeout.add_seconds (SECONDS_BETWEEN_REFRESHES, check_seconds_between_refreshes);
+    }
+
+    private bool check_seconds_between_refreshes () {
+        /* One cache update a day, keeps the doctor away! */
+        var seconds_since_last_refresh = new DateTime.now_utc ().difference (last_firmware_check) / GLib.TimeSpan.SECOND;
+        if (seconds_since_last_refresh >= SECONDS_BETWEEN_REFRESHES) {
+            last_firmware_check = new DateTime.now_utc ();
+            firmware_settings.set_int64 ("last-refresh-time", last_firmware_check.to_unix ());
+
+            refresh_firmware_updates.begin ();
+        }
+
+        return Source.CONTINUE;
+    }
+
+    private async void refresh_firmware_updates () {
+        yield firmware_update_notification_mutex.lock ();
```

Comment:
You're locking `firmware_update_notification_mutex` and unlocking `update_notification_mutex`, so this will result in a deadlock

Reply:
Thanks a lot! That was maybe a little much copypasta  😅